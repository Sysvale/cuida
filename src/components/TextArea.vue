<template>
	<div>
		<CdsBaseInput
			ref="baseInput"
			v-model="internalValue"
			v-bind="{...$attrs, ...props}"
			type="textarea"
			:support-link-url="supportLinkUrl"
			:support-link="supportLink"
			:floating-label="floatingLabel"
			@click="emitClick"
			@change="emitChange"
			@focus="emitFocus"
			@blur="emitBlur"
			@keydown="emitKeydown"
		/>
	</div>
</template>

<script setup>
import { ref, watch, useTemplateRef } from 'vue';
import {
	nativeEvents,
	nativeEmits,
} from '../utils/composables/useComponentEmits.js';
import CdsBaseInput from './BaseInput.vue';

const baseInputRef = useTemplateRef('baseInput');

const model = defineModel('modelValue', {
	type: String,
});

const props = defineProps({
	/**
	 * Especifica a label do textarea.
	 */
	label: {
		type: String,
		default: 'Label',
	},
	/**
	* Exibe asterisco que indica campo obrigatório (Obs: Não faz a validação)
	*/
	required: {
		type: Boolean,
		default: false,
	},
	/**
	 * Especifica o placeholder do textarea.
	 */
	placeholder: {
		type: String,
		default: 'Digite aqui a descrição',
	},
	/**
	 * Desabilita o input do textarea.
	 */
	disabled: {
		type: Boolean,
		default: false,
	},
	/**
	 * Especifica o estado do textarea. As opções são 'default', 'valid', 'invalid' e 'loading'.
	 */
	state: {
		type: String,
		default: 'default',
	},
	/**
	 * Especifica a mensagem de erro que será exibida, caso o estado do textarea seja inválido.
	 */
	errorMessage: {
		type: String,
		default: 'Valor inválido',
	},
	/**
	* Especifica mensagem de auxílio.
	*/
	supportingText: {
		type: [String, Array],
		default: '',
	},
	/**
	* Controla a exibição e o conteúdo do link de suporte exibido ao lado da label.
	*/
	supportLink: {
		type: String,
		default: '',
	},
	/**
	* Define a url a ser acessada no clique do link de suporte.
	*/
	supportLinkUrl: {
		type: String,
		default: 'https://cuida.framer.wiki/',
	},
	/**
	 * Especifica se a largura do textarea deve ser fluida.
	 */
	fluid: {
		type: Boolean,
		default: false,
		required: false,
	},
	/**
	* Quando true, o v-model é atualizado com o evento `change` no lugar do `input`.
	*/
	lazy: {
		type: Boolean,
		default: false,
	},
	/**
	* Define exibição e texto do tooltip do input
	*/
	tooltip: {
		type: String,
		default: null,
	},
	/**
	* Especifica ícone do tooltip do TextInput.
	*/
	tooltipIcon: {
		type: String,
		default: 'info-outline',
	},
	/**
	* Define o tipo do input, se true será um input adaptado para o mobile
	*/
	floatingLabel: {
		type: Boolean,
		default: false,
	},
});

const emits = defineEmits({
	...nativeEvents
});

const internalValue = ref(model.value);
const { emitClick, emitChange, emitFocus, emitBlur, emitKeydown } = nativeEmits(emits);

watch(model, (newValue, oldValue) => {
	if (newValue !== oldValue) {
		internalValue.value = newValue;
	}
}, { immediate: true });

watch(internalValue, (value) => {
	model.value = value;
});

/* EXPOSE */
defineExpose({
	componentRef: baseInputRef.value?.componentRef,
	isFocused: baseInputRef.value?.isFocused,
	focus: () => baseInputRef.value?.focus(),
	blur: () => baseInputRef.value?.blur(),
	clear: () => baseInputRef.value?.clear(),
	select: () => baseInputRef.value?.select(),
});
</script>
