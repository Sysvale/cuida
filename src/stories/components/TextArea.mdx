import { Meta, Story, Props, Controls, Canvas, Preview } from '@storybook/blocks';
import TextArea from '../../components/TextArea.vue';
import * as TextAreaStories from './TextArea.stories';

<Meta of={TextAreaStories} />

export const Template = (args) => ({
	components: { CdsTextArea: TextArea },
	setup() {
		return { args };
	},
	data() {
		return {
			value: '',
		};
	},
	template: /*html*/ `
		<cds-text-area
			v-bind="args" 
			v-model="value"
		/>
	`,
	updated() {
		console.info('%cv-model: ', 'color: #2C70CD;', this.value);
	},
});

# TextArea

### TextAreas permitem que os usuários insiram um texto longo em uma interface.

***

<br />

## Quando usar:

* Necessitar de um campo de texto para inserir informações grandes. Exemplo: Descrição de produto ou mensagens.

<br />

## Quando não usar:

* For necessário informar apenas um dado específico ao campo de texto.

<br />

## Preview

<Canvas>
  <Story of={TextAreaStories.TextArea} />
</Canvas>

<Controls />
