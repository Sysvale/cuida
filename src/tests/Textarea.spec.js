import { describe, test, expect } from 'vitest';
import { mount } from '@vue/test-utils';
import TextArea from '../components/TextArea.vue';
import CdsBaseInput from '../components/BaseInput.vue';

describe('TextArea', () => {
	test('renders correctly', () => {
		const wrapper = mount(TextArea, {
			props: {
				id: 'text-area',
			},
		});

		expect(wrapper.html()).toMatchSnapshot();
	});

	test('renders correctly with default props', () => {
		const wrapper = mount(TextArea, {
			props: {
				id: 'text-area',
			},
		});

		expect(wrapper.findComponent(CdsBaseInput).exists()).toBe(true);
		expect(wrapper.findComponent(CdsBaseInput).props('label')).toBe('Label');
		expect(wrapper.findComponent(CdsBaseInput).props('placeholder')).toBe('Digite aqui a descrição');
		expect(wrapper.findComponent(CdsBaseInput).props('disabled')).toBe(false);
		expect(wrapper.findComponent(CdsBaseInput).props('required')).toBe(false);
		expect(wrapper.findComponent(CdsBaseInput).props('fluid')).toBe(false);
		expect(wrapper.findComponent(CdsBaseInput).props('floatingLabel')).toBe(false);
	});

	test('renders with a required indicator', () => {
		const wrapper = mount(TextArea, {
			props: {
				id: 'text-area',
				required: true,
			},
		});

		expect(wrapper.findComponent(CdsBaseInput).props('required')).toBe(true);
	});

	test('renders with disabled attribute when disabled prop is true', () => {
		const wrapper = mount(TextArea, {
			props: {
				id: 'text-area',
				disabled: true,
			},
		});

		expect(wrapper.findComponent(CdsBaseInput).props('disabled')).toBe(true);
	});

	test('does not render with disabled attribute when disabled prop is false', () => {
		const wrapper = mount(TextArea, {
			props: {
				id: 'text-area',
				disabled: false,
			},
		});

		expect(wrapper.findComponent(CdsBaseInput).props('disabled')).toBe(false);
	});

	test('applies fluid class and styles when fluid prop is true', () => {
		const wrapper = mount(TextArea, {
			props: {
				id: 'text-area',
				fluid: true,
			},
		});

		expect(wrapper.findComponent(CdsBaseInput).props('fluid')).toBe(true);
	});

	test('renders with a custom label', () => {
		const wrapper = mount(TextArea, {
			props: {
				id: 'text-area',
				label: 'Custom Label',
			},
		});

		expect(wrapper.findComponent(CdsBaseInput).props('label')).toBe('Custom Label');
	});

	test('renders with a support link', () => {
		const wrapper = mount(TextArea, {
			props: {
				id: 'text-area',
				supportLink: 'Support Link',
				supportLinkUrl: 'https://example.com',
			},
		});

		expect(wrapper.findComponent(CdsBaseInput).props('supportLink')).toBe('Support Link');
		expect(wrapper.findComponent(CdsBaseInput).props('supportLinkUrl')).toBe('https://example.com');
	});

	test('renders with a floating label when floatingLabel prop is true', () => {
		const wrapper = mount(TextArea, {
			props: {
				id: 'text-area',
				floatingLabel: true,
			},
		});

		expect(wrapper.findComponent(CdsBaseInput).props('floatingLabel')).toBe(true);
	});

	test('updates the model value when input changes', async () => {
		const wrapper = mount(TextArea, {
			props: {
				id: 'text-area',
			},
		});

		const baseInput = wrapper.findComponent(CdsBaseInput);
		await baseInput.setValue('new value');
		expect(wrapper.vm.internalValue).toBe('new value');
	});
});
