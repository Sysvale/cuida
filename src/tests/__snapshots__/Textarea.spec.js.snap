// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`TextArea > renders correctly 1`] = `
"<div id="text-area">
  <div data-v-82c69faf="">
    <!--v-if--><label data-v-c1eb01da="" data-v-82c69faf="" class="label"><span data-v-c1eb01da="" class="label__content">Label <!--v-if--><!--v-if--></span>
      <!--v-if-->
    </label>
    <div data-v-82c69faf="" class="base-input base-input--default">
      <!--v-if--><textarea data-v-82c69faf="" id="cds-base-input-textarea-text-area" placeholder="Digite aqui a descriÃ§Ã£o" class="base-input__field" type="textarea"></textarea>
      <!--v-if-->
      <!--v-if-->
    </div>
    <!--v-if-->
    <!--v-if-->
  </div>
</div>"
`;
